rs/src/parse/relations/join.rs

```diff
@@ -38,10 +38,24 @@ pub fn parse_join_rel(x: &substrait::JoinRel, y: &mut context::Context) -> diagn
     }
 
     // Parse join expression.
-    let join_expression =
-        proto_boxed_required_field!(x, y, expression, expressions::parse_predicate)
-            .1
-            .unwrap_or_default();
+    let (join_expression_node, opt_join_expression) =
+        proto_boxed_required_field!(x, y, expression, expressions::parse_predicate);
+    let join_expression = opt_join_expression.unwrap_or_default();
+
+    // If the join expression is a literal it must be True.
+    if let expressions::Expression::Literal(ref literal) = join_expression {
+        if !matches!(
+            literal.value(),
+            expressions::literals::LiteralValue::Boolean(true)
+        ) {
+            diagnostic!(
+                y,
+                Error,
+                ExpressionIllegalLiteralValue,
+                "literal join expression value must be true"
+            );
+        }
+    }
```

Comment:
> literal join expression value must be true

Does it? I mean, yeah, to get a plan that makes any kind of sense, sure, but I don't think it is or should be illegal to do it, just like how `if false { ... }` isn't illegal. Errors and warnings are very strictly only dealt out to plans that are actually illegal, because the validity verdict is derived from it.

If this is something you really want to detect, you could emit a comment, or throw an info diagnostic in (I guess) the `Redundant` class (the JoinRel becomes kind of redundant at that point, since it would always emit zero rows, or in case of an anti join emit the LHS unchanged, I think).

Reply:
I agree with you, but I'm just following the [spec](https://github.com/substrait-io/substrait/blob/e68fdc049c1219a1269225456072d7548c83f297/site/docs/relations/logical_relations.md?plain=1#L199) here.

Reply:
All I see is

> Can be the literal True.

That doesn't mean it can't be the literal false, or any other expression. It just clarifies that it's legal to turn a JoinRel into a full cross product this way, despite the expression itself being required.